src/CodeGen_PTX_Dev.cpp

```diff
@@ -447,13 +448,13 @@ void CodeGen_PTX_Dev::codegen_vector_reduce(const VectorReduce *op, const Expr &
         {VectorReduce::Add, 4, i32(widening_mul(wild_u8x, wild_i8x)), "dp4a_u32_s32"},
         {VectorReduce::Add, 4, u32(widening_mul(wild_u8x, wild_u8x)), "dp4a_u32_u32"},
         {VectorReduce::Add, 4, widening_mul(wild_i16x, wild_i16x), "dp2a_s32_s32", Pattern::NarrowOp1},
-        {VectorReduce::Add, 4, widening_mul(wild_i16x, wild_u16x), "dp2a_s32_u32", Pattern::NarrowOp1},
+        {VectorReduce::Add, 4, widening_mul(wild_i16x, wild_i16x), "dp2a_s32_u32", Pattern::NarrowUnsignedOp1},
```

Comment:
Is the second arg here signed or unsigned?

Reply:
It's unsigned, widened to a signed integer.

Reply:
Then why are both args wild_i16x?

Reply:
When widening a uint8, it can be widened to either a uint16 or an int16. This is a pretty common issue, see e.g. https://github.com/halide/Halide/blob/master/src/HexagonOptimize.cpp#L534-L535 (on master).

`widening_mul` eliminates a lot of this issue, but it doesn't fully eliminate it in the mixed precision case. I suppose we could generalize `widening_mul` further to strip widening casts "unevenly", so this pattern would be `widening_mul(wild_i16x, wild_u8x)`...

Reply:
If we generalized it to support mixed precision, we'd have to be pretty careful about constants. We'd probably need to always narrow constants to the narrowest type they can represent.

Reply:
I realized a bit better way to handle this, using the overloaded intrinsics.